### DISCUSSION TITLE
DuckDB analytics tools like pg_stats in Postgres?

### DISCUSSION THREAD

**Discussion by david99s** [duckdb/duckdb#2400] [Q&A] [answered]
Hey, 
i was wondering if DuckDB contains some analyze tools/tables like Postgres does, e.g. pg_stats, pg_stat_statements?
Best Regards
David

**Comment by Mytherin** [comment, marked as answer]
Stats can be obtained per column through the `stats` function, e.g.:

```sql
CREATE TABLE tbl AS SELECT col, col::VARCHAR col_varchar FROM range(100) tbl(col);
SELECT stats(col), stats(col_varchar) FROM tbl LIMIT 1;
┌────────────────────────────────────┬──────────────────────────────────────────────────────────────────────────────┐
│             stats(col)             │                              stats(col_varchar)                              │
├────────────────────────────────────┼──────────────────────────────────────────────────────────────────────────────┤
│ [Min: 0, Max: 99][Has Null: false] │ [Min: 0, Max: 99, Has Unicode: false, Max String Length: 2][Has Null: false] │
└────────────────────────────────────┴──────────────────────────────────────────────────────────────────────────────┘

```

If you want to get an overview of what a table looks like you can also use the `SUMMARIZE` command, but these statistics are computed at run-time.

```sql
CREATE TABLE tbl AS SELECT col, col::VARCHAR col_varchar FROM range(100) tbl(col);
SUMMARIZE tbl;
┌─────────────┬─────────────┬─────┬─────┬───────────────┬──────┬────────────────────┬──────┬──────┬──────┬───────┬─────────────────┐
│ column_name │ column_type │ min │ max │ approx_unique │ avg  │        std         │ q25  │ q50  │ q75  │ count │ null_percentage │
├─────────────┼─────────────┼─────┼─────┼───────────────┼──────┼────────────────────┼──────┼──────┼──────┼───────┼─────────────────┤
│ col         │ BIGINT      │ 0   │ 99  │ 99            │ 49.5 │ 29.011491975882016 │ 24   │ 49   │ 74   │ 100   │ 0.0%            │
│ col_varchar │ VARCHAR     │ 0   │ 99  │ 100           │ NULL │ NULL               │ NULL │ NULL │ NULL │ 100   │ 0.0%            │
└─────────────┴─────────────┴─────┴─────┴───────────────┴──────┴────────────────────┴──────┴──────┴──────┴───────┴─────────────────┘
```

**Comment by david99s** [comment]
And for analyzing how the answer of the query get calculated? meaning which scans and joins DuckDb is using for this query? "Explain..." in postgres. I am really sorry to bother you but i can't find these things in your documentation on your website.

**Reply by derhuerst** [reply]
https://duckdb.org/docs/sql/statements/profiling

**Comment by Mytherin** [comment]
`stats` will be computed at bind-time, `summarize` requires a full scan of the table.

**Comment by david99s** [comment]
Thank you for your explanation. The question was actually asked in general. When i am executing the query e.g."explain select l_orderkey from lineitem" i only get "physical plan" as output but i am looking for an output like "seq scan..."

**Comment by Mytherin** [comment]
This seems to work as intended for me:

```sql
D call dbgen(sf=0.1);
D explain select l_orderkey from lineitem;

┌─────────────────────────────┐
│┌───────────────────────────┐│
││       Physical Plan       ││
│└───────────────────────────┘│
└─────────────────────────────┘
┌───────────────────────────┐
│          SEQ_SCAN         │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│          lineitem         │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│         l_orderkey        │
└───────────────────────────┘   
```

You can also use `PRAGMA explain_output='all'` to view multiple phases of the logical plan:

```sql
D pragma explain_output='all';
D explain select l_orderkey from lineitem;

┌─────────────────────────────┐
│┌───────────────────────────┐│
││ Unoptimized Logical Plan  ││
│└───────────────────────────┘│
└─────────────────────────────┘
┌───────────────────────────┐
│         PROJECTION        │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│         l_orderkey        │
└─────────────┬─────────────┘                             
┌─────────────┴─────────────┐
│          SEQ_SCAN         │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│          lineitem         │
└───────────────────────────┘                             

┌─────────────────────────────┐
│┌───────────────────────────┐│
││  Optimized Logical Plan   ││
│└───────────────────────────┘│
└─────────────────────────────┘
┌───────────────────────────┐
│         PROJECTION        │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│         l_orderkey        │
└─────────────┬─────────────┘                             
┌─────────────┴─────────────┐
│          SEQ_SCAN         │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│          lineitem         │
└───────────────────────────┘                             

┌─────────────────────────────┐
│┌───────────────────────────┐│
││       Physical Plan       ││
│└───────────────────────────┘│
└─────────────────────────────┘
┌───────────────────────────┐
│          SEQ_SCAN         │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│          lineitem         │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│         l_orderkey        │
└───────────────────────────┘  
```

**Reply by david99s** [reply]
Okay now its working for me either, thanks a lot. Does DuckDB support several commands like "explain costs", "explain analyze"? Because it does not show the costs/ execution time?

**Comment by Mytherin** [comment]
Those are not supported yet, but you you can run `PRAGMA enable_profiling` which prints out execution time of each individual operator after execution of a query. Actually a good point that we should support `EXPLAIN ANALYZE`! I will open an issue for that.

**Reply by david99s** [reply]
Okay thank you very much. I am currently working on my bachelor thesis in which i am working in Postgres and DuckDb to evaluate the influence of schema tuning (horizontal/vertical partitioning) for the performance of queries. I hope that it's not a problem for you if i keep continuing asking questions in the next time regarding the understanding of DuckDB and its inner mechanism. Thank you!

**Comment by david99s** [comment]
How do I have to work with this pragma command. I am coding on Java, so i don't know how to simply run this command. How should it be connected to the actual query? 
`String z = "explain select l_shipdate from orders join lineitem on o_orderkey=l_orderkey where o_orderkey=235671";
 ResultSet rs2 = b.executeQuery(z);
 ResultSet h =b.executeQuery("pragma enable_profiling");`

But iterating thorugh the resultSet h to get the information about the execution time does not work for me

**Reply by Mytherin** [reply]
When you enable the pragma profiling output of subsequent queries is printed to the terminal, the result of the query is still returned as normal. You can also use ‘PRAGMA profiling_output='test.json'’ to write output to a file instead, or use ‘SELECT * FROM pragma_last_profiling_output()’ to fetch the last profiling output as a query result.

**Reply by david99s** [reply]
`ResultSet rs3 = b.executeQuery("select * from pragma_last_profiling_output()");
if(rs3!=null){
System.out.println("true");
while(rs3.next()){
                    System.out.println("hi");
                    String time=rs3.getString(1);
                    String time2=rs3.getString(2);
                    System.out.println(time);
                    System.out.println(time2);
                }`
It says the Resultset is not null, but i cannot iterate through its output. Not even the "hi" gets printed

**Comment by aborruso** [comment]
@Mytherin how to create a table starting from SUMMARIZE?

It's not `create table s AS SUMMARIZE select * from  input. Parquet;`

Thank you

**Reply by Mytherin** [reply]
That is not possible currently because `SUMMARIZE` is a special top level statement. We should definitely allow for this, though.

**Reply by aborruso** [reply]
What about a feature request?

**Reply by aborruso** [reply]
> That is not possible currently because `SUMMARIZE` is a special top level statement. We should definitely allow for this, though.

My goal in example is to select all fields whose null values are greater than 83%. Moreover the NULL field is a VARCHAR and not a FLOAT.
I think it should became a FLOAT

**Comment by vlulla** [comment]
Related to @aborruso's question about creating table from `SUMMARIZE`, is there any plan to allow using `SUMMARIZE` as a from item? So, for instance: `select * from (summarize mytable) where column_name like 'se%' and column_type='VARCHAR'` would be very helpful in finding columns that have some invalid values (N/A, NA, Not Applicable) which changed columns from float/double to varchar. I have run across these annoyances surprisingly frequently and would find this feature quite useful. Regardless of what you decide, I love using duckdb! Thanks in advance!
